pkg/controller/rsm2/replica_util.go

```diff
@@ -0,0 +1,486 @@
+		replicas := int32(1)
+		if instance.Replicas != nil {
+			replicas = *instance.Replicas
+		}
+		replicasInTemplates += replicas
+
+		if instance.Name != nil {
+			if instance.Replicas != nil && *instance.Replicas > 1 {
+				names = fmt.Sprintf("%s%s,", names, *instance.Name)
+			}
+		}
+	}
+	// sum of spec.templates[*].replicas should not greater than spec.replicas
+	if replicasInTemplates > *rsm.Spec.Replicas {
+		return fmt.Errorf("total replicas in instances(%d) should not greater than replicas in spec(%d)", replicasInTemplates, *rsm.Spec.Replicas)
+	}
+
+	// instance.replicas should be nil or 1 if instance.name set
+	if len(names) > 0 {
+		return fmt.Errorf("replicas should be empty or no more than 1 if name set, instance names: %s", names)
+	}
+
+	return nil
+}
+
+func buildPodTemplateRevision(template *podTemplateSpecExt, parent *workloads.ReplicatedStateMachine) (string, error) {
+	// try to generate the same revision as if generated by sts controller
+	rsm := builder.NewReplicatedStateMachineBuilder(parent.Namespace, parent.Name).
+		SetUID(parent.UID).
+		AddAnnotationsInMap(parent.Annotations).
+		AddMatchLabelsInMap(parent.Labels).
+		SetTemplate(template.PodTemplateSpec).
+		GetObject()
+
+	cr, err := NewRevision(rsm)
+	if err != nil {
+		return "", err
+	}
+	return cr.Labels[ControllerRevisionHashLabel], nil
+}
+
+func buildReplicaTemplateGroups(rsm *workloads.ReplicatedStateMachine) map[string][]*podTemplateSpecExt {
+	var podTemplates []*podTemplateSpecExt
+	var replicasInTemplates int32
+	envConfigName := rsm1.GetEnvConfigMapName(rsm.Name)
+	defaultTemplate := rsm1.BuildPodTemplate(rsm, envConfigName)
+	buildPodTemplateExt := func(replicas int32) *podTemplateSpecExt {
+		var claims []corev1.PersistentVolumeClaim
+		for _, template := range rsm.Spec.VolumeClaimTemplates {
+			claims = append(claims, *template.DeepCopy())
+		}
+		return &podTemplateSpecExt{
+			Replicas:             replicas,
+			PodTemplateSpec:      *defaultTemplate.DeepCopy(),
+			VolumeClaimTemplates: claims,
+		}
+	}
+	for _, instance := range rsm.Spec.Instances {
+		replicas := int32(1)
+		if instance.Replicas != nil {
+			replicas = *instance.Replicas
+		}
+		template := buildPodTemplateExt(replicas)
+		applyInstanceTemplate(instance, template)
+		podTemplates = append(podTemplates, template)
+		replicasInTemplates += template.Replicas
+	}
+	if replicasInTemplates < *rsm.Spec.Replicas {
+		template := buildPodTemplateExt(*rsm.Spec.Replicas - replicasInTemplates)
+		podTemplates = append(podTemplates, template)
+	}
+	// set the default name generator and namespace
+	for _, template := range podTemplates {
+		if template.GenerateName == "" {
+			template.GenerateName = rsm.Name
+		}
+		template.Namespace = rsm.Namespace
+	}
+
+	// group the pod templates by template.Name if set or by template.GenerateName
+	replicaTemplateGroups := make(map[string][]*podTemplateSpecExt)
+	for _, template := range podTemplates {
+		name := template.Name
```

Comment:
template.name must be unique. otherwise will be override

Reply:
duplicate template.name is an invalid request and will be refused by the rsm2 controller.